### PR TITLE
Provide switch to allow remote connections for gradle daemon

### PR DESCRIPTION
**Pull request by kxepal**
Any of the checked boxes below indicate that I took action:

- [X] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).
- [X] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [X] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:

- [X] Before submitting a pull request, I started a discussion on the
[Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev)
or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [X] I wrote a design document. A design document can be
brief but explains the use case or problem you are trying to solve,
touches on the planned implementation approach as well as the test cases
that verify the behavior. Optimally, design documents should be submitted
as a separate pull request. [Samples](https://github.com/gradle/gradle/tree/master/design-docs)
can be found in the Gradle GitHub repository. Please let us know if you need help with
creating the design document. We are happy to help!
- [X] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
I ran a build on your local machine via the command
`./gradlew quickCheck <impacted-subproject>:check`.
- [X] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

This fixes common issue on FreeBSD jails when loopback interface
provides no IP addresses due to security reasons.

GRADLE-3121

### CODE REVIEW

**Comment by kxepal** [comment]
CLA had been just submitted. I was expected more paper work. Nice service!

File: subprojects/docs/src/docs/userguide/buildEnvironment.xml [hunk 1]
```diff
@@ -45,6 +45,14 @@
                 </listitem>
             </varlistentry>
             <varlistentry>
+                <term><literal>org.gradle.daemon.allowremote</literal></term>
+                <listitem><para>When set to <literal>true</literal> the Gradle
+                    daemon can accept connection not jsut from local addresses
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/buildEnvironment.xml]: Typo: `jsut`

Comment (kxepal) [inline subprojects/docs/src/docs/userguide/buildEnvironment.xml]: Oops, fixed. Thanks!

**Comment by wolfs** [comment]
Thank you for your contribution.

Wouldn't it be better if the daemon would detect on its own that there cannot be a connection on a local network interface and then allow remote connections? Do you know how your situation can be detected? The fix should either go in `DaemonTcpServerConnector` or `InetAddressFactory.init` depending on this is an issue with the daemon only or if sematically the ip-address you are referring to is a local address.

**Comment by kxepal** [comment]
@wolfs 

> Wouldn't it be better if the daemon would detect on its own that there cannot be a connection on a local network interface and then allow remote connections? Do you know how your situation can be detected?

The case that FreeBSD jail falls is that during counting available local/remote addresses, gradle fails to find any available local ones and [computes fallback address](https://github.com/gradle/gradle/blob/master/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/InetAddressFactory.java#L203-L207) to use. 

Technically, it's possible to not do that fallback, but turn on allowRemote flag, but I'm not feels good on this because it's about the same story as with database, that bind by default to `0.0.0.0` without any auth setup: this may cause security or else risks without giving developers any note about, because this behaviour change happens automagically and implicitly. I would leave such decision (which iface to bind on) to the userland.

> The fix should either go in DaemonTcpServerConnector or InetAddressFactory.init depending on this is an issue with the daemon only or if sematically the ip-address you are referring to is a local address.

Changes for DaemonTcpServerConnector is not enough to fix the problem because we also have to fix MessageHubBackedServer to follow the same behaviour. Otherwise `gradle test` will fail with the same issue and it doesn't uses Daemon here.

Another idea is to not pass `allowRemote` to Receiver, but let Receiver read SystemProperties (or else) and figure out should it accept connections from remote or not. But this will require some couple of tests to rewrite.

**Comment by wolfs** [comment]
The daemon validates that they pass the correct key and they can only get that from the daemon registry on the filesystem, but we don’t think we want to change anything in the non-daemon code (i.e. relaxing the fallback, etc). We think we just want some logic to determine if we’re in that fallback scenario and set allowRemote accordingly.
About the MessagingHubBackedServer: I understand that we run into the same problem here. I will investigate if there is some security in place, but I think that there is not. The question is wether this is an issue or not since we are talking about short lived processes here.

**Comment by kxepal** [comment]
@wolfs 
Thanks for the feedback. I'm not sure about any security things there as well, but my logic is that if there is allowRemote flag, then it was added with some reason and since it's false by default, there were some concerns about to not accept remotes by the same default. Otherwise, proposed auto-allowRemote set breaks that logic, that's my point of care.

**Comment by kxepal** [comment]
Digging into history, seems like @adammurdoch [added](https://github.com/gradle/gradle/commit/360890d60554d3bdd3202cf14863f5389f5c622f#diff-ecf7daf738a8deb2c97963672276bf67R107) that flag with reason to have better tests coverage without intentions to make it being usable from outside. So, gradle always had used local addresses without being accessible from remote.

**Comment by wolfs** [comment]
Sorry for the late response.

We had a discussion internally and we decided we are fine with setting both allowRemote flags to true in both locations. Please adjust your Pull request accordingly. Thank you.

**Comment by kxepal** [comment]
@wolfs 
Done with corrected the commit message. Should this change be also noted somewhere?

**Comment by wolfs** [comment]
If you could add a section to the release notes, this would be great. Do you know if the old behavior is documented somewhere? I did not find anything.

**Comment by kxepal** [comment]
@wolfs 
No, I didn't found any references of current behaviour as well ):
Updated PR with release notices addition. Could you check if I put changes to the right place?

P.S. Also, would it be possible to backport this patch to 2.x branch?

**Comment by wolfs** [comment]
@kxepal: Regarding the release notes: Please add a new section just before promoted features.

Since 2.14 will be the last release for the 2.x series and we are already in the RC-phase we cannot backport this change to 2.14. A 3.0 milestone will be available this week.

**Comment by kxepal** [comment]
@wolfs 
I'm a bit confused. How to name that section?

> Since 2.14 will be the last release for the 2.x series and we are already in the RC-phase we cannot backport this change to 2.14. A 3.0 milestone will be available this week.

Ok, good to know! Thanks.

**Comment by wolfs** [comment]
After thinking a bit more about this it is not necessary to add anything to the release notes. The Issue fixed will be added automatically and this will contain sufficient information for this change.

**Comment by kxepal** [comment]
Ok, I reverted changes in notes.md.

**Comment by wolfs** [comment]
Thank you for your contribution. Your changes have been merged and will be included in Gradle 3.0.

**Comment by wolfs** [comment]
Internal review: https://code-review.gradle.org/cru/REVIEW-6069

**Comment by kxepal** [comment]
Great! Thanks! (:

**Comment by wolfs** [comment]
The final fix was different to your PR: I changed InetAddressFactory to use a better fallback, which should work in your situation. Could you try if it fixes your issue?

The original change also had the impact that all Daemon connections used a non-loopback network interface which can have a serious performance impact.

**Comment by kxepal** [comment]
@wolfs 
Oh! Sorry for that! Actually, I didn't noticed any serious/notable delays with the fix. Probably, because of jail specifics. Will check in few hours.

**Comment by big-guy** [comment]
@kxepal we'll be putting out a RC with this fix soon-ish, so if you could give a 3.0-nightly a try (https://services.gradle.org/distributions-snapshots) and let us know if you have any trouble, that would be very helpful.

**Comment by kxepal** [comment]
Oh, sorry, I completely forgot about! ): Hold on...

**Comment by kxepal** [comment]
Hm...I cannot have my project built with the latest 3.0 snapshot so far:
```
./gradlew build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
No compatible daemons found:
  - a busy daemon can't run this build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
No compatible daemons found:
  - 2 busy daemons can't run this build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
No compatible daemons found:
  - 3 busy daemons can't run this build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
No compatible daemons found:
  - 4 busy daemons can't run this build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
No compatible daemons found:
  - 5 busy daemons can't run this build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
No compatible daemons found:
  - 6 busy daemons can't run this build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
No compatible daemons found:
  - 7 busy daemons can't run this build
Starting a new Gradle Daemon for this build: subsequent builds will be faster
```
and so it keeps going. 2.13 works like a charm. I'll backport the patch onto it to test since the code base for that scope looks similar.

**Comment by kxepal** [comment]
Ah, nvm, I pick the wrong snapshot. `gradle-3.0-20160720000031+0000` works great! Thanks a lot, @wolfs!

**Pull request merged**